File: src/main/java/com/snackgame/server/member/domain/Name.java
package com.snackgame.server.member.domain;

import java.util.Objects;
import java.util.regex.Pattern;

import javax.persistence.Column;
import javax.persistence.Embeddable;

import com.snackgame.server.member.exception.EmptyNameException;
import com.snackgame.server.member.exception.NameLengthException;

import lombok.AccessLevel;
import lombok.NoArgsConstructor;

@Embeddable
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Name {

    private static final Pattern NUMBERED_PATTERN = Pattern.compile(".+_\\d+");
    @Column(name = "name")
    private String string;

    public Name(String string) {
        validateNotNull(string);
        validateLengthOf(string);
        this.string = string;
    }

    private void validateNotNull(String string) {
        if (Objects.isNull(string)) {
            throw new EmptyNameException();
        }
    }

    private void validateLengthOf(String string) {
        if (string.length() < 2 || string.length() > 16) {
            throw new NameLengthException();
        }
    }

    public Name nextAvailable() {
        if (NUMBERED_PATTERN.matcher(string).matches()) {
            int underscoreIndex = string.lastIndexOf('_');
            long currentNumber = Long.parseLong(string.substring(underscoreIndex + 1));
            return with(currentNumber + 1);
        }
        return with(2);
    }

    private Name with(long suffix) {
        int underscoreIndex = string.lastIndexOf('_');
        if (underscoreIndex > 0) {
            return new Name(string.substring(0, underscoreIndex) + '_' + suffix);
        }
        return new Name(string + '_' + suffix);
    }

    public String getString() {
        return string;
    }
}

File: src/main/java/com/snackgame/server/member/domain/Name.kt
package com.snackgame.server.member.domain

import com.snackgame.server.member.exception.NameLengthException
import java.util.regex.Pattern
import javax.persistence.Column
import javax.persistence.Embeddable

@Embeddable
class Name(
    @Column(name = "name")
    val string: String
) {

    init {
        validateLengthOf(string)
    }

    private fun validateLengthOf(string: String) {
        if (string.length < 2 || string.length > 13) {
            throw NameLengthException()
        }
    }

    fun nextAvailable(): Name {
        if (Pattern.matches(NUMBERED_PATTERN, string)) {
            val currentNumber = string.substringAfterLast('_').toLong()
            return Name("${string.substringBeforeLast('_')}_${currentNumber + 1}")
        }
        return Name(string + "_2")
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Name

        return string == other.string
    }

    override fun hashCode(): Int {
        return string.hashCode()
    }

    companion object {
        private const val NUMBERED_PATTERN = ".+_\\d+"
    }
}
